### PR TITLE
[READY TO MERGE] Improve Portfolio Optimisation Documentation (Part-1)

### PR DESCRIPTION
**Pull request by aditya1702**
# Description
The current documentation for portfolio optimisation is a bit lacking in the explanations about the algorithms and in general how to use the code. This PR adds more readability to it with better explanations and links to references.

### CODE REVIEW

**Comment by Jackal08** [comment]
This PR is of a very high priority.

File: docs/source/portfolio_optimisation/hierarchical_risk_parity.rst [hunk 1]
```diff
@@ -39,6 +80,41 @@ Implementation
...
+        * :py:mod:`asset_prices`: Dataframe/matrix of historical raw asset prices **indexed by date**.
+        * :py:mod:`asset_returns`: Dataframe/matrix of historical asset returns. This will be a :math:`TxN` matrix where :math:`T` is the time-series and :math:`N` refers to the number of assets in the portfolio.
+        * :py:mod:`expected_asset_returns`: List of expected returns per asset i.e. the mean of historical asset returns. This refers to the parameter :math:`\mu` used in portfolio optimisation literature. For a portfolio of 5 assets, ``expected_asset_returns = [0.45, 0.56, 0.89, 1.34, 2.4]``.
+        * :py:mod:`covariance_matrix`: The covariance matrix of asset returns.
+
+
+.. tip::
+    |h4| Using Custom Distance Matrix |h4_|
+    The hierarchical clustering step in the algorithm uses a distance matrix to calculate the clusters and form the hierarchical
+    tree. By default, we use the distance matrix mentioned in the original paper,
```
Comment (PanPip) [inline docs/source/portfolio_optimisation/hierarchical_risk_parity.rst:97]: Original paper is included only in the docstring of the `HierarchicalRiskParity` class. Perhaps also add it as a Tip?

File: docs/source/portfolio_optimisation/hierarchical_risk_parity.rst [hunk 2]
```diff
@@ -39,6 +80,41 @@ Implementation
...
+
+    |h4| Constructing a Long/Short Portfolio |h4_|
+    By default, the weights are allocated assuming a long portfolio i.e. all the weights are positive. However, users can also opt
+    to have specific assets in their portfolio shorted, by specifying a custom list through the :py:mod:`side_weights` parameter.
+    For e.g. in a portfolio of 5 assets, if you want the 3rd and the 5th asset to be shorted, then simply pass a list containing
+    1s and -1s for long and short positions respectively. In the above case, ``side_weights = [1,1,-1,1,-1]``, where you
+    have -1 at the 3rd and 5th index. By default, a list of 1s will be initialised if no custom input is specified.
+
+    |h4| Different Linkage Methods |h4_|
+    HRP, by default, uses the single-linkage clustering algorithm. (See the tip under the HCAA algorithm for more details.)
```
Comment (PanPip) [inline docs/source/portfolio_optimisation/hierarchical_risk_parity.rst:114]: Perhaps, worth adding that the list of supported methods is `...` . And later - They're described in the tip under the HCAA algorithm.

File: docs/source/portfolio_optimisation/hierarchical_clustering_asset_allocation.rst [hunk 3]
```diff
@@ -10,20 +30,77 @@
...
+    .. image:: portfolio_optimisation_images/gap.png
+
+    |br|
+
+    |h3| **3. Top-Down Recursive Bisection** |h3_|
+    This is the step where weights for the clusters are calculated. If you are familiar with how the hierarchical risk parity
+    algorithm works, then you know this is similar to how HRP also allocates its weights. However, there is a fundamental
+    difference between the recursive bisections of the two algorithms.
+
+    .. image:: portfolio_optimisation_images/bisection.png
```
[the remark this answers is not in the document]
Comment (aditya1702) [reply docs/source/portfolio_optimisation/hierarchical_clustering_asset_allocation.rst:64]: I do not think they are needed. Also, most of them are taken from the papers themselves.

File: docs/source/portfolio_optimisation/hierarchical_risk_parity.rst (hunk 1 above)
Comment (aditya1702) [reply docs/source/portfolio_optimisation/hierarchical_risk_parity.rst:97]: Good idea

**Pull request merged**